Allow all configured http methods in the CORS preflight response

hamburml opened the pull request:
During cors preflight response should always return all configured http methods which are allowed. If none are configured allow everything.

https://github.com/quarkusio/quarkus/discussions/33155

hamburml committed:
During cors preflight response should always return all configured http methods which are allowed. If none are configured allow everything.

sberyozkin commented:
Hi @hamburml, before commiting to such a refactoring, can you please confirm that the browser rejects POST if you can configure the filter to allow GET ?

sberyozkin commented:
@hamburml Let me check it with the tests as well

sberyozkin commented:
@hamburml I just updated the title of the PR to have the bot happier :-)

hamburml commented:
> Hi @hamburml, before commiting to such a refactoring, can you please confirm that the browser rejects POST if you can configure the filter to allow GET and DELETE as per the discussion you opened ?

I am not sure but if I remember correctly the browser did not reject it when the status code of preflight was 200 (even when POST is not in preflight allowed-methods header). When the status code was 403 it did reject the POST request. I only have time in the evening to test it again and provide some screenshots.

Maybe it is a compatibility thing (https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests) that requests which are understood as "simple requests" always are carried out and the values inside the cors headers are ignored.

sberyozkin commented:
@hamburml Sure, let me work with your PR now, thanks for your help

hamburml commented:
By the way, I can very well understand not making this change because it could break stuff. I just think it is strange to disallow a http method but when it is used it works.

sberyozkin commented:
@hamburml We'll have to do it if it fixes the bug :-). I think overall the origin check offers a very strong protection as it will prevent things like stealing the session cookies etc, so when it comes to a method level control, one would always have for ex a stronger RBAC restriction for the methods with the side-effects. Nonetheless, we need to fix this issue as well. 
Cheers

sberyozkin committed:
Fix CORS tests

sberyozkin commented:
Hey @hamburml I've fixed the tests and made sure there is a dedicated error/log message related to the not allowed method. Thanks

sberyozkin commented:
Sorry, one more tweak related to the wildcard case is needed

stuartwdouglas commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -221,6 +221,11 @@ public void handle(RoutingContext event) {
                 response.setStatusMessage("CORS Rejected - Invalid origin");
                 response.end();
+            } else if (!allowsMethod) {
+                LOG.debug("Method is not allowed");
```
This is confusing a preflight request with the actual request itself. For the actual request you check the method, there will be no 'ACCESS_CONTROL_REQUEST_METHOD' header.

In general this existing code is not ideal. It is applying preflight checks for access control headers to ever request, which leads to this kind of confusion. 

This code needs to be refactored to:

- Split out the normal request and preflight code paths
- in the normal request code path check the actual method (and the headers), and return a 403 if there is a failure
- In the preflight path just return all configured methods and headers, there is not any need to actually parse the Access-Control-Request-Headers or Access-Control-Request-Methods header, all we need to check is if they are present or not.

sberyozkin replied:
Sure, thanks for the refactoring

stuartwdouglas committed:
Refactor CORS support

This refactoring make it clear the different betweeen preflight and
normal requests.

It also adds access control to normal requests such that they will be
blocked if an invalid header or method is present.

stuartwdouglas commented:
I have pushed some changes here that significantly refactor how this works.

stuartwdouglas commented on extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSHandlerTestWildcardOriginCase.java:
```diff
@@ -103,19 +103,4 @@ void corsInvalidSameOriginRequest5() {
-    @Test
-    @DisplayName("Returns false 'Access-Control-Allow-Credentials' header on matching origin '*'")
-    void corsMatchingOriginWithWildcard() {
-        String origin = "*";
```
* is not a valid value for the origin header.

stuartwdouglas commented on extensions/reactive-routes/deployment/src/test/java/io/quarkus/vertx/web/cors/CORSHandlerTestCase.java:
```diff
@@ -31,25 +31,23 @@ public void corsPreflightTestServlet() {
+        String methods = "POST";
         String headers = "X-Custom";
         given().header("Origin", origin)
-                .header("Access-Control-Request-Method", methods)
```
Note that these two headers are only valid for preflight (i.e. OPTIONS) requests. They should be ignored otherwise as they have no defined meaning outside an options request.

sberyozkin replied:
@stuartwdouglas `Origin` can be sent with simple requests, we've had the related issues

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -176,64 +161,91 @@ public void handle(RoutingContext event) {
-                }
+                return;
+            }
+            if (!allowsOrigin) {
```
@stuartwdouglas I think this check should be moved just above the block which sets `403` for the not allowed method, otherwise the not allowed method status message will override the not allowed origin status message which is more important I guess.

sberyozkin commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -21,6 +25,24 @@
 
 public class CORSFilter implements Handler<RoutingContext> {
 
+    static final Set<String> SIMPLE_HEADERS;
```
@stuartwdouglas Why does this CORS filter need to worry about these subset of well known headers and content types ? Not all the headers/media types are listed here, what is the plan ? They are not used anywhere in the filter code just yet and I'm not sure CORS filter should be doing the related checks for the simple requests...

sberyozkin replied:
Right, do you mean enforce that only those headers which are configured to be allowed, do make it during the simple request ? It can be hard to do though, as some headers like User-Agent, content negotiation ones, transfer encoding, etc can be added. Not sure we need to deal with it... For the preflight we have a list of allowed headers if needed

stuartwdouglas replied:
I need to delete this. These headers are called out in the fetch spec as being explicitly CORS safe, and I originally thought I would block requests based on allowed headers, but that is not actually how Access-Control-Allowed-Headers works. 

This specifies the headers that the JavaScript is allowed to set on the request, not the allowed headers for the request, as the browser will set significatly more headers than what is specified in the allowed headers.

sberyozkin replied:
Sure, makes sense, I can delete it for now and we can revisit

sberyozkin commented:
@stuartwdouglas Thanks for the refactoring, I have a few questions there, re the simple headers/media types, one of the test status codes, and minor code block move. I can tweak but we need to agree before I do it.
By the way, can we assume `GET, POST` (with space) works same way as `GET,POST` for all the browsers ?
thanks

hamburml commented on extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java:
```diff
@@ -98,59 +98,46 @@ public void corsPreflightTest() {
                 .auth().basic("user", "user")
                 .options("/test").then()
-                .statusCode(403)
+                .statusCode(200)
```
@sberyozkin @stuartwdouglas Yeah, when the cors preflight response statuscode is 200 but `Access-Control-Allow-Methods` does not contain the method which is inside `Access-Control-Request-Method` the browser was not blocking the actual request. If you like you can check it for yourself here https://github.com/hamburml/quarkus-cors-example.

stuartwdouglas replied:
This is a pre-flight options request. It is not the actual request. The important part is the value of the Access-Control-Allowed-Methods etc, not the return code.

If the response code was important then none of these response headers would be required, the server would just return a 200 or a 403 based on if it is allowed or not.

If you look at the fetch spec in particular this is called out here: https://fetch.spec.whatwg.org/#http-responses

`They can provide a static response. This can be helpful when working with caching intermediaries. A static response can both be successful and not successful depending on the [CORS request](https://fetch.spec.whatwg.org/#cors-request). This is okay.`

If the user has specified all the different properties in our CORS config we could just return a static response every time. Our response is dynamic because if the user has not specified a value we respond with what is requested in the HTTP request. 

We still return a 403 to the actual POST request. CORS will never block POST,GET or HEAD no matter the value of Access-Control-Allowed-Methods (basically because the JS can just create a form an submit the form and get a POST request that way). If you change your example to PUT rather than post and use this version of Quarkus you will see the browser blocking it.

hamburml replied:
Yes, I know that it is the preflight options request and not the actual request.

> CORS will never block POST,GET or HEAD no matter the value of Access-Control-Allowed-Methods

That is true. Cors it not a firewall so it can not block the request. But the browser can decide to not carry out the actual request and it uses the response of the cors preflight as decisive factors. And in my example this only worked when the cors preflight options request had a 403 status code. 

> just return a static response every time

Nah, in our aws api gateway we had a static response for cors options request. But the `Access-Control-Allow-Origin` should always be the origin of the request if you have more than one origin who is allowed to use this resouce.

> If you change your example to PUT rather than post and use this version of Quarkus you will see the browser blocking it.

Thanks, Ill try it out later. I simply do not understand why CORS never blocks POST,GET,HEAD if I can remove them from the allowed-methods list. It would be unambiguous to not allow POST,GET,HEAD in Access-Control-Allowed-Methods because it simply does not matter (because never blocked). Thx for your explanation.

stuartwdouglas replied:
It is because there are workarounds to invoking a POST/GET request. To do a GET request the JS can just create an <image> element with the source as the GET request to invoke. To do a POST request the JS can just create a FORM element and then call form.submit().

Blocking these requests would just create the illusion of security, as the JS could still invoke them.

hamburml replied:
oh wow, now i got it. Thx!

sberyozkin replied:
@stuartwdouglas Sure, but returning `403` would be consistent with how we handle mismatched Origins for the preflights. This filter did return `200` for mismatched origins for preflight requests but then we changed it to `403` after some discussions which you may recall (incl for it to be consistent with Vert.x)
 
I see, if we block the method for an actual POST/etc request anyway, then, for the preflight request, it does not really matter if it is 403 or not, but if 403 can prevent some browsers even trying to follow up with an actual request then it should be good.
 
Can we agree on `403` in case of preflight method or header mis-matches ?

stuartwdouglas replied:
Header mismatch is an absolute no-go, it's not how headers work. Access-Control-Allow-Headers tells the browser the headers that the JS can set on the request. The browser can and does add its own headers outside this set. 

Methods are also bad IMHO. If we return a 200 the browser can cache this information and use it to avoid further pre-flight requests. I think even returning a 403 on origin mismatch for preflight is potentially wrong, however the cache is not really an issue there so it is probably fine.

Like I said earlier the browser blocking on a 403 for incorrect method only adds the illusion of security.

sberyozkin replied:
@stuartwdouglas Makes sense now, thanks for the clarifications about the cache, sounds good

geoand commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -323,4 +330,61 @@ static boolean substringMatch(String str, int pos, String substring, boolean req
+        if (wildcardMethod) {
+            response.headers().set(HttpHeaders.ACCESS_CONTROL_ALLOW_METHODS, allowMethodsValue);
+        } else {
+            StringBuilder allowMethods = new StringBuilder();
```
Tiny gripe: We can provide a good estimate of the size here.

sberyozkin replied:
Hi @geoand, sorry, missed this comment, if the default is 16 then I guess it would be just right, ex, we have tests with 4 configured methods, `GET,POST,PUT,DELETE`, but these are 4 main methods, the users can just set a wildcard :-). So sometimes it can overflow by a few characters, but it is more likely than not 16 chars should be enough. Does it sound reasonable ?

geoand replied:
My point is that the we already know exactly what is going in to the builder, so we know the size.

In any case, it's a minor thing that can be addressed in a follow up

sberyozkin replied:
Sure, I think I see the point; I might've seen it yesterday now that you have explained it but I hesitated as I was not too keen on checking the already configured methods length... In any case thanks for helping out with the performance advice :-)

geoand commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:
```diff
@@ -172,73 +161,91 @@ public void handle(RoutingContext event) {
+
+        //normal request
+        //always set expose headers if present
+        if (exposedHeaders != null) {
+            response.headers().add(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, exposedHeaders);
+        }
+
+        if (!isConfiguredWithWildcard(corsConfig.exposedHeaders)) {
+            response.headers().set(HttpHeaders.ACCESS_CONTROL_EXPOSE_HEADERS, this.exposedHeaders);
+        }
```
I don't fully understand what's going here, first we are adding a header value and then we are setting it? I think these blocks need more code comments

sberyozkin replied:
Good catch, looks like a copy and paste problem, `exposedHeaders` will not be null only if it is not a wildcard config

sberyozkin commented:
@stuartwdouglas I did a couple of minor updates (removed the well known safe headers set for now, moved the allowedOrigins checks above the methods), I also rebased to pick up Georgios's fix to avoid seeing elytron security JDBC failures

sberyozkin reviewed (approved):
@hamburml Thanks for triggering this refactoring with your initial PR

@stuartwdouglas Thanks for your help as always, and for refactoring.

sberyozkin commented:
@stuartwdouglas I've lifted your change requests, all is green, I can merge it a bit later, after everyone has another look

hamburml commented:
Happy to help and thanks @stuartwdouglas for your explanations :)
